src/MainWindow.vala

```diff
@@ -1338,8 +1337,8 @@ namespace Terminal {
             string[] zooms = {};
 
             Application.saved_state.set_double ("zoom", current_terminal.font_scale);
-
-            if (Granite.Services.System.history_is_enabled () &&
+            if (tabs_were_restored &&  // Do not save tabs if window opened without restoring to avoid overwriting
+                Granite.Services.System.history_is_enabled () &&
```

Comment:
I believe the agreement in the files PR was that we always save the last closed window state. Not the first opened one.